_posts/2016-09-30-subtype-typeclasses.md

```diff
@@ -0,0 +1,402 @@
+danger of implicit ambiguity. However, for that very reason, having a
+`Monad[F]` no longer implies having a `Functor[F]`. Not currently
+anyway. What we can do is use implicit conversions to re-encode the
+hierarchy.
+
+```scala
+implicit def applicativeIsFunctor[F[_]: Applicative]: Functor[F] =
+  implicitly[Applicative[F]].functor
+
+implicit def traverseIsFunctor[F[_]: Traverse]: Functor[F] =
+  implicitly[Traverse[F]].functor
+```
+
+But now we're back to square one.
+
+```scala
+// Syntax for Functor
+implicit class FunctorOps[F[_], A](fa: F[A])(implicit F: Functor[F]) {
+  def map[B](f: A => B): F[B] = F.map(fa)(f)
+}
+```
+
+```scala
+def foo[F[_]: Applicative: Traverse]: F[Int] =
+  implicitly[Applicative[F]].pure(10).map(identity)
+// <console>:19: error: value map is not a member of type parameter F[Int]
+//          implicitly[Applicative[F]].pure(10).map(identity)
+//                                              ^
+// <console>:19: error: missing argument list for method identity in object Predef
+// Unapplied methods are only converted to functions when a function type is expected.
+// You can make this conversion explicit by writing `identity _` or `identity(_)` instead of `identity`.
+//          implicitly[Applicative[F]].pure(10).map(identity)
+//                                                  ^
+```
+
+Since both implicit conversions have same [priority][implicits], the compiler
+doesn't know which one to pick. **However**, Scala has mechanisms for
+[prioritizing implicits][implicits] which solves the problem.
+
+```scala
+object Prioritized { // needed for tut, irrelevant to demonstration
+  trait Functor[F[_]] {
+    def map[A, B](fa: F[A])(f: A => B): F[B]
+  }
+
+  // Prioritize implicit conversions - Functor only for brevity
+  trait FunctorConversions1 {
+    implicit def applicativeIsFunctor[F[_]: Applicative]: Functor[F] =
+      implicitly[Applicative[F]].functor
+  }
+
+  trait FunctorConversions0 extends FunctorConversions1 {
+    implicit def traverseIsFunctor[F[_]: Traverse]: Functor[F] =
+      implicitly[Traverse[F]].functor
+  }
+
+  object Functor extends FunctorConversions0
+
+  trait Applicative[F[_]] {
+    def functor: Functor[F]
+
+    def pure[A](a: A): F[A]
+    def map2[A, B, C](fa: F[A], fb: F[B])(f: (A, B) => C): F[C]
+  }
+
+  // Definition elided for space
+  trait Traverse[F[_]] { def functor: Functor[F] }
+
+  // Syntax for Functor
+  implicit class FunctorOps[F[_], A](fa: F[A])(implicit F: Functor[F]) {
+    def map[B](f: A => B): F[B] = F.map(fa)(f)
+  }
+
+  def foo[F[_]: Applicative: Traverse]: F[Int] = {
+    implicitly[Applicative[F]] // we have Applicative
+    implicitly[Traverse[F]]    // we have Traverse
+    implicitly[Functor[F]]     // we also have Functor!
+
+    // and we have syntax!
+    implicitly[Applicative[F]].pure(10).map(identity)
+  }
+}
+```
+
+The trick is to use trait linearization which gives the effect of implicits
```

Comment:
I don't think this can be called a result of trait linearization, especially because it works with classes' subclasses too.